onnx/backend/test/case/node/__init__.py

```diff
@@ -103,27 +103,19 @@ def function_testcase_helper(node, input_types, name):  # type: (NodeProto, List
     return node_list
 
 
-def _extract_value_info(input, name, ele_type=None):  # type: (Union[List[Any], np.ndarray, None], Text, np.dtype) -> onnx.ValueInfoProto
-    if isinstance(input, list):
-        # TODO: Account for recursive sequence case. Right now, this function supports
-        # Sequences of Tensors.
-        return onnx.helper.make_tensor_sequence_value_info(
-            name=name,
-            elem_type=ele_type if ele_type else onnx.mapping.NP_TYPE_TO_TENSOR_TYPE[input[0].dtype],
-            shape=None
-        )
-    elif input is None:
-        if ele_type is None:
-            raise NotImplementedError("extract_value for optional tensor with no data type not "
-                                      "supported.")
-        else:
-            tensor_type_proto = onnx.helper.make_tensor_type_proto(ele_type, shape=None)
-            optional_type_proto = onnx.helper.make_optional_type_proto(tensor_type_proto)
-            return onnx.helper.make_value_info(name=name, type_proto=optional_type_proto)
-    return onnx.helper.make_tensor_value_info(
-        name=name,
-        elem_type=ele_type if ele_type else onnx.mapping.NP_TYPE_TO_TENSOR_TYPE[input.dtype],
-        shape=input.shape)
+def _extract_value_info(input, name, type_proto=None):  # type: (Union[List[Any], np.ndarray, None], Text, TypeProto) -> onnx.ValueInfoProto
+    if type_proto is None:
+        if isinstance(input, list):
+            elem_type = onnx.mapping.NP_TYPE_TO_TENSOR_TYPE[input[0].dtype]
```

Comment:
please add this comment back :

 # TODO: Account for recursive sequence case. Right now, this function supports
 # Sequences of Tensors.

we are still only supporting Sequence of tensors

Reply:
In the case where type_proto is None there is no else block to set type_proto... meaning if control does not enter 
if isinstance(input, list) or elif isinstance(input, np.ndarray) or np.isscalar(input) then type_proto will be None when the control reaches line 118.

Reply:
how will this work in the case when optional proto holds None value? type_proto will be None?

Reply:
Agree. Better to have an assertion/check if the value is None.

Reply:
@askhade I don't think we need the comment back. The function now does support sequence of sequence or other recursive types if the type_proto for such type is provided. So users can create this kind of type_proto and then use _extract_value_info to create the value info.

Reply:
@askhade In the case where optional_proto holds None, we rely on the user to provide a corresponding type_proto. For a None tensor, the type_proto has elem_type (depending on the None tensor data type).
I'll add a check to fail if both input and type_proto are None.